include/nlohmann/detail/macro_scope.hpp

```diff
@@ -230,6 +230,46 @@
         e = ((it != std::end(m)) ? it : std::begin(m))->first;                                  \
     }
 
+#if (defined(__cpp_exceptions) || defined(__EXCEPTIONS) || defined(_CPPUNWIND)) && !defined(JSON_NOEXCEPTION)
```

Comment:
If this only works when `JSON_NOEXCEPTION` isn't defined, then that should be noted in the docs.  However, I think you can just use `JSON_THROW` instead of `throw` and use the normal exception override behavior.

Reply:
agree, will note it in the docs. I think there are some arguments for not using `JSON_NOEXCEPTION`:

- nested macros don't work (the preprocessor doesn't expand a macro inside another one)
- it can lead to silent and hard to debug erros if `JSON_NOEXCEPTION` is defined and the macro is used -> what should we do in that case?
- the macro itself doens't make any sense if `JSON_NOEXCEPTION` is defined (hence better not define it in the first place.

Reply:
ok, very interesting. Thanks for the godbolt link! Now I'm confused, my compiler (gcc 12, fedora 37) doesn't like that:
```
/home/pablo/workspace/json/tests/src/unit-conversions.cpp:1665:9:   required from here
/home/pablo/workspace/json/include/nlohmann/detail/macro_scope.hpp:249:23: error: ‘JSON_THROW’ was not declared in this scope; did you mean ‘JSON_TRY’?
  249 |             JSON_THROW(nlohmann::detail::type_error::create(302,                                                                   \
      |             ~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  250 |                     nlohmann::detail::concat("can't serialize enum value ", std::to_string(static_cast<int>(e)),              \
      |                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  251 |                                              " of enum '", #ENUM_TYPE, "' to json. ",                                          \
      |                                              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  252 |                                              "Did you list the value in NLOHMANN_JSON_SERIALIZE_ENUM_STRICT?"), &j));          \
      |                                              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/pablo/workspace/json/tests/src/unit-conversions.cpp:1618:1: note: in expansion of macro ‘NLOHMANN_JSON_SERIALIZE_ENUM_STRICT’
 1618 | NLOHMANN_JSON_SERIALIZE_ENUM_STRICT(game,
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

I wonder what I'm doing wrong

Reply:
My first thought was that `JSON_THROW` wasn't defined yet, but that's not the case.  It appears right before that macro, and it's not possible to get there without it being defined.  I'm not sure why gcc would choke on that.  The only other thing I can think of is that somehow you have something equivalent to `#define JSON_THROW JSON_THROW`.  Maybe `#define JSON_THROW_USER JSON_THROW`.  Can you reproduce this on compiler explorer?

Reply:
I'm really clueless, would it be possible for you to try the branch out with the proposed changes locally? Looks like my preprocessor doesn't see `JSON_THROW` and lets it through to the compiler...